apollo-router/src/query_planner/bridge_query_planner.rs

```diff
@@ -533,6 +577,65 @@ impl BridgeQueryPlanner {
             }
         }
     }
+
+    // todo move this somewhere else? somewhere where it's possible to do fuzzing?
+    fn generate_apollo_reporting_signature(doc: &ParsedDocument) -> String {
+        // todo this implementation is temporary and needs a lot of work
+
+        let definitions = doc.ast.definitions.clone();
+        let operation_def = definitions[0].as_operation_definition().unwrap().clone();
+
+        let operation_body = operation_def.to_string();
+        let operation_name = operation_def.name.as_ref().unwrap().to_string();
+        // println!("operation_body: {}", operation_body);
+        // println!("operation_name: {}", operation_name);
+
+        let whitespace_regex = regex::Regex::new(r"\s+").unwrap();
+        let stripped_body = whitespace_regex.replace_all(&operation_body, " ").to_string();
+
+        format!("# {}\n{}", operation_name, stripped_body)
+    }
+
+    // todo move this somewhere else? somewhere where it's possible to do fuzzing?
```

Comment:
We do publish https://crates.io/crates/apollo-router for Rust plugins, so its dependencies need to be published as well. Would https://github.com/apollographql/federation-next/ / https://crates.io/crates/apollo-federation be a good place for this?

Reply:
That follows the pattern of the JS libraries too. The functionality I'm porting in this PR is in the https://github.com/apollographql/apollo-utils repo, not the federation repo.

Reply:
@lrlna I don't think I can put this code into a separate crate. It has dependencies on some classes/structs in the router crate like `ParsedDocument` and `ReferencedFieldsForType` so if I used it in its own crate I think there would be a circular dependency between apollo-router and this new interop crate?

If I just move the code to a public function in the apollo-router crate, can I set up the fuzzer to call that?

Reply:
Can you use `apollo_compiler::ExecutableDocument` as input directly, instead of `ParsedDocument`?

`ReferencedFieldsForType` currently lives in router-bridge, but I think post replatforming it would make more sense if it is defined together with the Rust-based sig+ref code. (probably can only do this when we no longer rely on the JS implementation at all.)

Reply:
Thank you! That works perfectly. I was stuck on thinking that I needed to use the executable document rather than the ast::Document so I was stuck with ParsedDocument, but I didn't realise that the ExecutableDocument part is in apollo-rs. I've had to restructure my tests a bit but it looks like it's working without using anything from the router crate now so I should be good to go with making it a separate crate!

Reply:
@lrlna @goto-bus-stop sorry I'm running into another issue. I've moved the code to a new internal crate and I'm trying to write a fuzzer for it (code is [here](https://github.com/apollographql/router/pull/4796/files#diff-bf2f89b92b60adc030312b587c8775a6d3cfbe8813a5af95c849471a2da634e5)). I've managed to use the new rust code to generate the signature and referenced fields, but when I try to instantiate a Planner object to compare that with the JS version, I keep running into an error I don't understand. When I run the code
```
let result = Planner::<serde_json::Value>::new(ts.to_string(), QueryPlannerConfig::default()).await;
```
Where `ts` is a string that represents a valid Schema, I just get this error:
```
thread '<unnamed>' panicked at /Users/nickmarsh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/router-bridge-0.5.16+v2.7.1/src/worker.rs:101:38:
called `Result::unwrap()` on an `Err` value: TypeError: Error parsing args at position 1: serde_v8 error: invalid type; expected: object, got: string
```
If either of you has any thoughts on how I could fix this I'd appreciate it.

Reply:
> If I just move the code to a public function in the apollo-router crate, can I set up the fuzzer to call that?

That would work too.
>I also started trying to write the fuzzer in apollo-rs-fuzz

having this work in a fuzz_target in the router is fine imo

I don't think moving this to `apollo-rs-fuzz` is going to fix this error. That looks like a serde parsing error (json serialisation). What's the type system that `apollo-smith` generated for you here that causes this (very unlikely culprit, but just to double check)?

Reply:
I'm not exactly sure what you mean but I've got the schema and operation being generated as Strings and I'm trying to create the planner this way:
```

    let runtime = Runtime::new().unwrap();
    let _planner = runtime.block_on(planner(&schema_str));

...

async fn planner(ts: &str) -> Planner<serde_json::Value> {
    let result =
        Planner::<serde_json::Value>::new(ts.to_string(), QueryPlannerConfig::default()).await;
    println!("======= got past Planner::new =======");
   ...
}
```

And it never gets past the Planner::new. It looks like you're out of office but I'll try to get someone to help me out during observability friends tonight.

Reply:
putting it in a separate crate will increase the release load, because now we have to publish that crate, without any real benefit because making it public in that crate or in apollo-router would be the same from the fuzzer's point of view. So let's put it in apollo-router. As @goto-bus-stop said it can be stored in `apollo_router::_private` if we don't want to commit on a stable API there

Reply:
It took me a little while to figure it out how to do that but I've done it now and I ran into a very odd dependency issue. When I try to run the fuzzer using
```
cargo +nightly fuzz run apollo_router_studio_interop
```

I get the error
```
error[E0635]: unknown feature `stdsimd`
  --> /Users/nickmarsh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ahash-0.8.6/src/lib.rs:99:42
   |
99 | #![cfg_attr(feature = "stdsimd", feature(stdsimd))]
```

[Apparently](https://users.rust-lang.org/t/error-e0635-unknown-feature-stdsimd/106445) this is because of a transitive dependency to `ahash` and we'd need to upgrade from v0.8.6 to v0.8.7, but the dependency chain is very complicated. Any thoughts on how I can resolve this?

<details>
  <summary>Dependency chain</summary>
```
❯ cargo tree -i ahash
ahash v0.8.6
├── hashbrown v0.14.1
│   ├── dashmap v5.5.3
│   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   └── serial_test v3.0.0
│   │       [dev-dependencies]
│   │       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   ├── indexmap v2.2.3
│   │   ├── apollo-compiler v1.0.0-beta.13
│   │   │   ├── apollo-federation v0.0.8
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   ├── apollo-federation v0.0.8 (*)
│   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   ├── h2 v0.3.24
│   │   │   ├── aws-smithy-runtime v1.1.6
│   │   │   │   ├── aws-config v1.1.6
│   │   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── aws-sdk-sso v1.14.0
│   │   │   │   │   └── aws-config v1.1.6 (*)
│   │   │   │   ├── aws-sdk-ssooidc v1.14.0
│   │   │   │   │   └── aws-config v1.1.6 (*)
│   │   │   │   └── aws-sdk-sts v1.14.0
│   │   │   │       └── aws-config v1.1.6 (*)
│   │   │   ├── hyper v0.14.28
│   │   │   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── aws-config v1.1.6 (*)
│   │   │   │   ├── aws-smithy-runtime v1.1.6 (*)
│   │   │   │   ├── axum v0.6.20
│   │   │   │   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │   └── tonic v0.9.2
│   │   │   │   │       ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │       ├── opentelemetry-otlp v0.13.0
│   │   │   │   │       │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │       └── opentelemetry-proto v0.3.0
│   │   │   │   │           └── opentelemetry-otlp v0.13.0 (*)
│   │   │   │   │   [dev-dependencies]
│   │   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── hyper-rustls v0.24.2
│   │   │   │   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │   ├── aws-smithy-runtime v1.1.6 (*)
│   │   │   │   │   └── reqwest v0.11.24
│   │   │   │   │       ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │       ├── opentelemetry-datadog v0.8.0
│   │   │   │   │       │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │       ├── opentelemetry-http v0.9.0
│   │   │   │   │       │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │       │   ├── opentelemetry-datadog v0.8.0 (*)
│   │   │   │   │       │   ├── opentelemetry-jaeger v0.19.0
│   │   │   │   │       │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │       │   ├── opentelemetry-otlp v0.13.0 (*)
│   │   │   │   │       │   └── opentelemetry-zipkin v0.18.0
│   │   │   │   │       │       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │       ├── opentelemetry-jaeger v0.19.0 (*)
│   │   │   │   │       ├── opentelemetry-otlp v0.13.0 (*)
│   │   │   │   │       └── opentelemetry-zipkin v0.18.0 (*)
│   │   │   │   │       [dev-dependencies]
│   │   │   │   │       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── hyper-timeout v0.4.1
│   │   │   │   │   └── tonic v0.9.2 (*)
│   │   │   │   ├── hyperlocal v0.8.0
│   │   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │   [dev-dependencies]
│   │   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── reqwest v0.11.24 (*)
│   │   │   │   ├── tonic v0.9.2 (*)
│   │   │   │   └── wiremock v0.5.22
│   │   │   │       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │       [dev-dependencies]
│   │   │   │       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── reqwest v0.11.24 (*)
│   │   │   └── tonic v0.9.2 (*)
│   │   ├── petgraph v0.6.4
│   │   │   ├── apollo-federation v0.0.8 (*)
│   │   │   └── daggy v0.8.0
│   │   │       └── test-span v0.7.0
│   │   │           [dev-dependencies]
│   │   │           └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   ├── serde_json v1.0.114
│   │   │   ├── access-json v0.1.0
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── apollo-federation v0.0.8 (*)
│   │   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── assert-json-diff v2.0.2
│   │   │   │   └── wiremock v0.5.22 (*)
│   │   │   ├── axum v0.6.20 (*)
│   │   │   ├── deno_core v0.200.0
│   │   │   │   ├── deno_console v0.115.0
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1
│   │   │   │   │       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │   [build-dependencies]
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   ├── deno_crypto v0.129.0
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   │   [build-dependencies]
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   ├── deno_url v0.115.0
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   │   [build-dependencies]
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   ├── deno_web v0.146.0
│   │   │   │   │   ├── deno_crypto v0.129.0 (*)
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   │   [build-dependencies]
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   ├── deno_webidl v0.115.0
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   │   [build-dependencies]
│   │   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   │   [build-dependencies]
│   │   │   │   └── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   ├── graphql_client v0.13.0
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── http-types v2.12.0
│   │   │   │   └── wiremock v0.5.22 (*)
│   │   │   ├── jsonpath-rust v0.3.5
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── jsonpath_lib v0.3.0
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── jsonschema v0.17.1
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── jsonwebtoken v9.2.0
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── opentelemetry-stdout v0.1.0
│   │   │   │   [dev-dependencies]
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── opentelemetry-zipkin v0.18.0 (*)
│   │   │   ├── opentelemetry_sdk v0.20.0
│   │   │   │   ├── opentelemetry v0.20.0
│   │   │   │   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │   ├── opentelemetry-datadog v0.8.0 (*)
│   │   │   │   │   ├── opentelemetry-jaeger v0.19.0 (*)
│   │   │   │   │   ├── opentelemetry-semantic-conventions v0.12.0
│   │   │   │   │   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │   │   ├── opentelemetry-datadog v0.8.0 (*)
│   │   │   │   │   │   ├── opentelemetry-jaeger v0.19.0 (*)
│   │   │   │   │   │   ├── opentelemetry-otlp v0.13.0 (*)
│   │   │   │   │   │   └── opentelemetry-zipkin v0.18.0 (*)
│   │   │   │   │   ├── opentelemetry-zipkin v0.18.0 (*)
│   │   │   │   │   └── tracing-opentelemetry v0.21.0
│   │   │   │   │       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   │   [dev-dependencies]
│   │   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── opentelemetry-otlp v0.13.0 (*)
│   │   │   │   ├── opentelemetry-prometheus v0.13.0
│   │   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── opentelemetry-proto v0.3.0 (*)
│   │   │   │   ├── opentelemetry-stdout v0.1.0 (*)
│   │   │   │   └── tracing-opentelemetry v0.21.0 (*)
│   │   │   ├── proteus v0.5.0
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── reqwest v0.11.24 (*)
│   │   │   ├── router-bridge v0.5.16+v2.7.1 (*)
│   │   │   ├── schemars v0.8.16
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── serde_json_bytes v0.2.2
│   │   │   │   ├── apollo-compiler v1.0.0-beta.13 (*)
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   ├── sourcemap v6.4.1
│   │   │   │   └── deno_core v0.200.0 (*)
│   │   │   ├── test-span v0.7.0 (*)
│   │   │   ├── tracing-subscriber v0.3.18
│   │   │   │   ├── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── test-log v0.2.14
│   │   │   │   │   [dev-dependencies]
│   │   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   ├── test-span v0.7.0 (*)
│   │   │   │   ├── tracing-opentelemetry v0.21.0 (*)
│   │   │   │   └── tracing-test v0.2.4
│   │   │   │       [dev-dependencies]
│   │   │   │       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   │   [dev-dependencies]
│   │   │   │   └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
│   │   │   └── wiremock v0.5.22 (*)
│   │   └── serde_json_bytes v0.2.2 (*)
│   └── lru v0.12.2
│       └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
├── jsonschema v0.17.1 (*)
└── rhai v1.17.1
    └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
    [dev-dependencies]
    └── apollo-router v1.43.0 (/Users/nickmarsh/src/router/apollo-router)
```
</details>

Reply:
@bonnici It looks like it will work with just updating the deps here:
```
cargo update -p ahash
cargo update -p curve25519-dalek@4.0.0
```
(somehow that other dependency has a similar issue)

Reply:
Thank you for that but unfortunately I ran into yet another strange issue I have no idea how to fix. The `cargo update -p` commands got me past the unknown feature `stdsimd` error but now I'm getting an even stranger linker error:

```
error: linking with `cc` failed: exit status: 1
  |
  = note: env -u IPHONEOS_DEPLOYMENT_TARGET -u TVOS_DEPLOYMENT_TARGET LC_ALL="C" PATH="/Users/nickmarsh/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/bin:/opt/homebrew/opt/curl/bin:/Users/nickmarsh/google-cloud-sdk/bin:/Users/nickmarsh/.volta/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/nickmarsh/.cargo/bin:/Users/nickmarsh/.rover/bin:/Users/nickmarsh/Library/Application Support/JetBrains/Toolbox/scripts" VSLANG="1033" ZERO_AR_DATE="1" "cc" "-arch" "arm64" "/var/folders/q2/y4vh72_9197819gxrzgwtx3r0000gn/T/rustcQOdT9N/symbols.o" "-Wl,-rpath" "-Xlinker" "/Users/nickmarsh/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib" "-lrustc-nightly_rt.asan" "/Users/nickmarsh/src/router/target/aarch64-apple-darwin/release/deps/apollo_router_studio_interop-38c9c488e24b25f8.apollo_router_studio_interop.a0f57017173ec7f8-cgu.0.rcgu.o" "/Users/nickmarsh/src/router/target/aarch64-apple-darwin/release/deps/apollo_router_studio_interop-38c9c488e24b25f8.1euljvum1py0sn1w.rcgu.o" "-L" 

< huge list of folders gods here >

"-lc++" "-framework" "SystemConfiguration" "-framework" "Security" "-framework" "CoreFoundation" "-framework" "Security" "-liconv" "-lSystem" "-lc" "-lm" "-L" "/Users/nickmarsh/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib" "-o" "/Users/nickmarsh/src/router/target/aarch64-apple-darwin/release/deps/apollo_router_studio_interop-38c9c488e24b25f8" "-nodefaultlibs"
  = note: ld: warning: search path '/Users/nickmarsh/src/router/target/aarch64-apple-darwin/release/build/libz-ng-sys-2ace1440088f968b/out/lib64' not found
          ld: initializer pointer has no target in '/Users/nickmarsh/src/router/target/aarch64-apple-darwin/release/deps/libproteus-6844fdb36340bdae.rlib[3](proteus-6844fdb36340bdae.proteus.9e4a919a603b00d1-cgu.0.rcgu.o)'
          clang: error: linker command failed with exit code 1 (use -v to see invocation)


error: could not compile `router-fuzz` (bin "apollo_router_studio_interop") due to 1 previous error
Error: failed to build fuzz script: ASAN_OPTIONS="detect_odr_violation=0" RUSTFLAGS="-Cpasses=sancov-module -Cllvm-args=-sanitizer-coverage-level=4 -Cllvm-args=-sanitizer-coverage-inline-8bit-counters -Cllvm-args=-sanitizer-coverage-pc-table -Cllvm-args=-sanitizer-coverage-trace-compares --cfg fuzzing -Clink-dead-code -Zsanitizer=address -Cdebug-assertions -C codegen-units=1" "cargo" "build" "--manifest-path" "/Users/nickmarsh/src/router/fuzz/Cargo.toml" "--target" "aarch64-apple-darwin" "--release" "--config" "profile.release.debug=true" "--bin" "apollo_router_studio_interop"
```

The fuzz build also started taking ages to finish (15+ minutes) and using up most of my free memory. Any thoughts?

Reply:
it looks like the fuzzing framework is trying to do too much, while the router's code is too big. So we need to change gears. I left suggestions in the PR that will make it build and run quickly, I just tested. The nice thing with that solution is that we won't need to reexport the module.
You also need to remove this line: https://github.com/apollographql/router/blob/181e8f72a29a4742e65c62b8175e7264ddafd304/fuzz/fuzz_targets/federation.rs#L2
Meanwhile I have a PR up to build the fuzzers in CI and make sure we don't regress there https://github.com/apollographql/router/pull/4925

Reply:
Thank you! I had to make a few other changes but it looks like it's all working now.